sky/serve/autoscalers.py

```diff
@@ -192,49 +236,210 @@ def evaluate_scaling(
+
+    This autoscaler uses spot instances to save cost.
+    """
+
+    NAME: Optional[str] = 'SpotRequestRateAutoscaler'
+
+    def __init__(self, spec: 'service_spec.SkyServiceSpec') -> None:
+        super().__init__(spec)
+        self.spot_placer: 'spot_policy.SpotPlacer' = (
+            spot_policy.SpotPlacer.from_spec(spec))
+
+    def handle_active_history(self, history: List[str]) -> None:
+        for zone in history:
+            self.spot_placer.handle_active(zone)
+
+    def handle_preemption_history(self, history: List[str]) -> None:
+        for zone in history:
+            self.spot_placer.handle_preemption(zone)
+
+    def _get_spot_resources_override_dict(self) -> Dict[str, Any]:
+        # We have checked before any_of can only be used to
+        # specify multiple zones, regions and clouds.
+        return {
+            'use_spot': True,
+            'spot_recovery': None,
+            'region': None,
+            'any_of': None
+        }
+
+    def _scale_spot_instances(
+        self, num_launched_spot: int,
+        launched_replica_infos: List['replica_managers.ReplicaInfo']
+    ) -> List[AutoscalerDecision]:
+        """Compare the current launched number of spot instances with the
+        number of spot instances to provision (including overprovisioning),
+        scale up or down accordingly.
+        """
+        scaling_options = []
+        num_to_provision = (self.target_num_replicas + self.num_overprovision)
+
+        # Scale spot instances.
+        if num_launched_spot < num_to_provision:
+            # Not enough spot instances, scale up.
+            # Consult spot_placer for the zone to launch spot instance.
+            num_spot_to_scale_up = num_to_provision - num_launched_spot
+            zones = self.spot_placer.select(launched_replica_infos,
+                                            num_spot_to_scale_up)
+            assert len(zones) == num_spot_to_scale_up
+            for spot_idx in range(num_spot_to_scale_up):
+                spot_override = self._get_spot_resources_override_dict()
+                zone = zones[spot_idx]
+                spot_override.update({'zone': zone})
+                logger.info(f'Chosen zone {zone} with {self.spot_placer}')
+                scaling_options.append(
+                    AutoscalerDecision(AutoscalerDecisionOperator.SCALE_UP,
+                                       target=spot_override))
+        elif num_launched_spot > num_to_provision:
+            # Too many spot instances, scale down.
+            # Get the replica to scale down with get_replica_ids_to_scale_down
+            num_spot_to_scale_down = num_launched_spot - num_to_provision
+            for replica_id in (
+                    RequestRateAutoscaler.get_replica_ids_to_scale_down(
+                        num_spot_to_scale_down, [
+                            info for info in launched_replica_infos
+                            if info.is_spot
```

Comment:
nit: Could we use `filter` to simplify the code? same for other similar code in the PR